airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/destination/buffered_stream_consumer/BufferedStreamConsumer.java

```diff
@@ -24,93 +24,101 @@
+ * as the onClose successfully commits any messages to the raw table.
+ * </p>
+ *
+ * <p>
+ * All other message types are ignored.
+ * </p>
+ */
 public class BufferedStreamConsumer extends FailureTrackingAirbyteMessageConsumer {
 
   private static final Logger LOGGER = LoggerFactory.getLogger(BufferedStreamConsumer.class);
-  private static final long THREAD_DELAY_MILLIS = 500L;
-
-  private static final long GRACEFUL_SHUTDOWN_MINUTES = 5L;
-  private static final int MIN_RECORDS = 500;
   private static final int BATCH_SIZE = 10000;
 
   private final VoidCallable onStart;
   private final RecordWriter recordWriter;
   private final CheckedConsumer<Boolean, Exception> onClose;
   private final Set<AirbyteStreamNameNamespacePair> pairs;
-  private final Map<AirbyteStreamNameNamespacePair, CloseableQueue<byte[]>> pairToWriteBuffer;
-  private final ScheduledExecutorService writerPool;
+  private final BlockingQueue<AirbyteMessage> queue;
   private final ConfiguredAirbyteCatalog catalog;
   private final CheckedFunction<String, Boolean, Exception> isValidRecord;
   private final Map<AirbyteStreamNameNamespacePair, Long> pairToIgnoredRecordCount;
+  private final Consumer<AirbyteStateMessage> checkpointConsumer;
 
   private boolean hasStarted;
+  private boolean hasClosed;
+
+  private AirbyteStreamNameNamespacePair lastRecordStream;
+  private AirbyteStateMessage lastCommittedState;
+  private AirbyteStateMessage pendingState;
 
   public BufferedStreamConsumer(VoidCallable onStart,
                                 RecordWriter recordWriter,
                                 CheckedConsumer<Boolean, Exception> onClose,
                                 ConfiguredAirbyteCatalog catalog,
-                                Set<AirbyteStreamNameNamespacePair> pairs,
                                 CheckedFunction<String, Boolean, Exception> isValidRecord) {
+    this(onStart, recordWriter, onClose, catalog, isValidRecord, (stateMessage) -> {});
+  }
+
+  // todo (cgardens) checkpointConsumer will become relevant once we start actually checkpointing.
+  @VisibleForTesting
+  BufferedStreamConsumer(VoidCallable onStart,
+                         RecordWriter recordWriter,
+                         CheckedConsumer<Boolean, Exception> onClose,
+                         ConfiguredAirbyteCatalog catalog,
+                         CheckedFunction<String, Boolean, Exception> isValidRecord,
+                         Consumer<AirbyteStateMessage> checkpointConsumer) {
+    this.checkpointConsumer = checkpointConsumer;
     this.hasStarted = false;
+    this.hasClosed = false;
     this.onStart = onStart;
     this.recordWriter = recordWriter;
     this.onClose = onClose;
     this.catalog = catalog;
-    this.pairs = pairs;
+    this.pairs = AirbyteStreamNameNamespacePair.fromConfiguredCatalog(catalog);
     this.isValidRecord = isValidRecord;
+    this.queue = new ArrayBlockingQueue<>(BATCH_SIZE);
```

Comment:
You actually don't need a blocking queue you since your producing and consuming in the same thread and in the same function. Just a regular array list would work here.